pkg/cli/sso.go

```diff
@@ -0,0 +1,216 @@
+			stdcli.StringFlag("client_id", "c", "client id"),
+			stdcli.StringFlag("client_secret", "s", "client secret"),
+			stdcli.StringFlag("issuer", "i", "issuer"),
+		},
+		Validate: stdcli.ArgsMax(0),
+	})
+
+	registerWithoutProvider("sso login", "authenticate with a console using sso", SsoLogin, stdcli.CommandOptions{
+		Validate: stdcli.ArgsMax(0),
+	})
+}
+
+func SsoConfigure(rack sdk.Interface, c *stdcli.Context) error {
+	var err error
+	provider := coalesce(c.String("provider"), os.Getenv("SSO_PROVIDER"))
+	if provider == "" {
+		c.Writef("SSO Provider: ")
+
+		provider, err = c.ReadSecret()
+		if err != nil {
+			return err
+		}
+
+		c.Writef("\n")
+	}
+
+	clientID := coalesce(c.String("client_id"), os.Getenv("SSO_CLIENT_ID"))
+	if clientID == "" {
+		c.Writef("SSO Client ID: ")
+
+		clientID, err = c.ReadSecret()
+		if err != nil {
+			return err
+		}
+
+		c.Writef("\n")
+	}
+
+	clientSecret := coalesce(c.String("client_secret"), os.Getenv("SSO_CLIENT_SECRET"))
+	if clientSecret == "" {
+		c.Writef("SSO Client Secret: ")
+
+		clientSecret, err = c.ReadSecret()
+		if err != nil {
+			return err
+		}
+
+		c.Writef("\n")
+	}
+
+	issuer := coalesce(c.String("issuer"), os.Getenv("SSO_ISSUER"))
+	if issuer == "" {
+		c.Writef("SSO ISSUER: ")
+
+		issuer, err = c.ReadSecret()
+		if err != nil {
+			return err
+		}
+
+		c.Writef("\n")
+	}
+
+	if err := c.SettingWrite("provider", provider); err != nil {
+		return err
+	}
+
+	if err := c.SettingWrite("client_id", clientID); err != nil {
+		return err
+	}
+
+	if err := c.SettingWrite("client_secret", clientSecret); err != nil {
+		return err
+	}
+
+	if err := c.SettingWrite("issuer", issuer); err != nil {
+		return err
+	}
+
+	return c.OK()
+}
+
+func SsoLogin(rack sdk.Interface, c *stdcli.Context) error {
+	provider, err := c.SettingRead("provider")
+	if err != nil {
+		return err
+	}
+
+	clientID, err := c.SettingRead("client_id")
+	if err != nil {
+		return err
+	}
+
+	clientSecret, err := c.SettingRead("client_secret")
+	if err != nil {
+		return err
+	}
+
+	issuer, err := c.SettingRead("issuer")
+	if err != nil {
+		return err
+	}
+
+	nonce, _ := sso.GenerateNonce()
+
+	p, err := sso.Initialize(provider, structs.SsoProviderOptions{
+		ClientID:     clientID,
+		ClientSecret: clientSecret,
+		Issuer:       issuer,
+		Scope:        "openid profile email",
+		State:        sso.GenerateState(),
+		Nonce:        nonce,
+	})
+
+	if err != nil {
+		return err
+	}
+
+	server := &http.Server{Addr: "/authorization-code/callback"}
+
+	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		authCodeCallbackHandler(w, r, server, c, p)
+	})
+
+	port := fmt.Sprintf(":%s", "8090")
+	l, err := net.Listen("tcp", port)
+	if err != nil {
+		fmt.Printf("snap: can't listen to port %s: %s\n", port, err)
+		os.Exit(1)
+	}
+
+	log.Println("Waiting for login... ")
+
+	time.Sleep(1 * time.Second)
+	sso.Openbrowser(p.RedirectPath())
+	time.Sleep(1 * time.Second)
+
+	server.Serve(l)
```

Comment:
what happens if the user don't finish the process? will the server be open? is it possible to add a timeout option? after X min turn it down